### ISSUE TITLE
New rule enforce a type with '*-of-type' rules

### ISSUE THREAD

**Issue by amk221** [stylelint/stylelint#4614]
> What is the problem you're trying to solve?

Often I come across code like:

```css
.foo:last-of-type {

}
```

and I'm left wondering, did the developer really mean last of type? Or would last child actually have been a better tool for the job? I can't know, without asking the developer, or investigating further by checking what type of element `.foo` is, and whether it has siblings of a different type.

An inexperienced developer may even think `.foo:last-of-type` means the last foo. Which it does not.

> What solution would you like to see?

I'm suggesting a rule that makes sure `-of-type` rules are always preceded with a type, e.g.

```css
p:last-of-type {

}

/* or even */

p.foo:last-of-type {

}
```

**Comment by jeddy3** [comment]
@amk221 Thanks for the suggestion and for using the template.

It's an interesting use case.

I think there are some related use cases too that are worth exploring.

If a user is disconnecting their styles from the underlying tags by using classes then I would suggest they use the [`selector-pseudo-class-blacklist` rule](https://stylelint.io/user-guide/rules/selector-pseudo-class-blacklist) to disallow the use of these [typed child-indexed pseudo-classes](https://drafts.csswg.org/selectors-4/#typed-child-index):

```json
{
  "rules": {
    "selector-pseudo-class-blacklist": ["/of-type$/"]
  }
}
```

If a user is connecting their styles to the underlying tags, then they may want to use the full capabilities of typed child-indexed pseudo-classes, which includes using the [implicit universal selector](https://developer.mozilla.org/en-US/docs/Web/CSS/:last-of-type#Nested_elements) (and limiting the scope of that with a class, e.g. `.foo:last-of-type {}`).

And then there's the use case you describe of allowing typed child-indexed pseudo-classes but disallowing the implicit universal selector. I think this use case is a great candidate for a [a plugin](https://stylelint.io/developer-guide/plugins) as I'm not sure ["generally useful" it is](https://stylelint.io/user-guide/rules/about#about-rules), even though it is valid.

I'd suggest naming the plugin `stylelint-pseudo-class-typed-child-indexed-no-implicit-universal` to [follow the conventions](https://stylelint.io/user-guide/rules/about#names). A bit long perhaps, but I believe it's an accurate description of what the plugin will do.

I've labelled as "needs: discussion" to see if anyone else has thoughts on this.

**Comment by jeddy3** [comment]
As there has been no further discussion. Closing the issue as the use case is best solved [as a plugin](https://stylelint.io/developer-guide/plugins).
